src/metabase/db/custom_migrations.clj

```diff
@@ -1033,3 +1033,28 @@
 (define-reversible-migration UnifyTimeColumnsType
   (unify-time-column-type! :up)
   (unify-time-column-type! :down))
+
+(define-reversible-migration CardRevisionAddType
+  (let [migrate! (fn [revision]
+                   (let [object    (json/parse-string (:object revision) keyword)
+                         new-object (-> object
+                                        (assoc :type (if (:dataset object)
+                                                       "model"
+                                                       "question")))]
+                     (t2/query {:update :revision
+                                :set    {:object (json/generate-string new-object)}
+                                :where  [:= :id (:id revision)]})))]
+    (run! migrate! (t2/reducible-query {:select [:*]
```

Comment:
Ouch these reducible migrations seem very expensive. Is it perhaps worth special casing the migration for databases that support JSON operations natively?

Reply:
This is a fair point. I think this will be the most expensive migration to date. On a large instance there can be 100,000s of these updates needed, and from memory this kind of update is prohibitively slow.

All our app DBs support JSON operations to implement this with SQL. I've tried updating JSON objects with SQL before ([PR](https://github.com/metabase/metabase/pull/27660/files)), the downside is it gets ugly very quickly for complex queries and it's a lot easier to make mistakes (but this is migration is much simpler than that PR).

I think you have a couple of options here:
1. migrate the cards in a `revision/revert-to-revision!` implementation for cards
2. make this migration fast enough using SQL in a way.

I lean towards 2 because any bugs with the migration are likely to be caught immediately, as opposed to 1.

Reply:
This is not the 1st time we've done it, and it seems we're ok with it; I've seen no failure reported for this kind of migration.

And also h2 does not support this out of the box, we'll have to inject our own UDF implementation like you did in #27660

Reply:
optimized for postgres and mysql in 70521d704c

Reply:
Awesome. I think H2 is fine to leave with the slow clojure version.

> This is not the 1st time we've done it

I meant this is the first time we will be migrating revisions for all cards. We have only migrated dashboard revisions from clojure before. Previous clojure migrations on card revisions filter out the majority of cards in the select statement.